# 01 Data Prep

## Imports

In [1]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns

import os
from pathlib import Path

## Dataset

In [2]:
base_ordner = Path().resolve().parent
raw_data_ordner = base_ordner / "data" / "raw"
proc_data_ordner = base_ordner / "data" / "processed"

In [3]:
df_raw = pd.read_csv(
    raw_data_ordner / "company_esg_financial_dataset.csv"
    )

## EDA

In [4]:
df_raw.shape

(11000, 16)

In [5]:
df_raw.head()

,CompanyID,CompanyName,Industry,Region,Year,Revenue,ProfitMargin,MarketCap,GrowthRate,ESG_Overall,ESG_Environmental,ESG_Social,ESG_Governance,CarbonEmissions,WaterUsage,EnergyConsumption
0,1,Company_1,Retail,Latin America,2015,459.2,6.0,337.5,NaN,57.0,60.7,33.5,76.8,35577.4,17788.7,71154.7
1,1,Company_1,Retail,Latin America,2016,473.8,4.6,366.6,3.2,56.7,58.9,32.8,78.5,37314.7,18657.4,74629.4
2,1,Company_1,Retail,Latin America,2017,564.9,5.2,313.4,19.2,56.5,57.6,34.0,77.8,45006.4,22503.2,90012.9
3,1,Company_1,Retail,Latin America,2018,558.4,4.3,283.0,-1.1,58.0,62.3,33.4,78.3,42650.1,21325.1,85300.2
4,1,Company_1,Retail,Latin America,2019,554.5,4.9,538.1,-0.7,56.6,63.7,30.0,76.1,41799.4,20899.7,83598.8


In [6]:
df_raw.tail()

,CompanyID,CompanyName,Industry,Region,Year,Revenue,ProfitMargin,MarketCap,GrowthRate,ESG_Overall,ESG_Environmental,ESG_Social,ESG_Governance,CarbonEmissions,WaterUsage,EnergyConsumption
10995,1000,Company_1000,Consumer Goods,Latin America,2021,3419.2,21.7,14416.7,0.4,52.2,67.2,46.3,43.1,498875.8,623594.7,2494378.8
10996,1000,Company_1000,Consumer Goods,Latin America,2022,3518.8,21.4,8078.9,2.9,51.9,68.7,44.7,42.3,506013.7,632517.1,2530068.4
10997,1000,Company_1000,Consumer Goods,Latin America,2023,3417.1,17.1,13263.4,-2.9,52.9,72.4,43.1,43.2,473561.5,591951.9,2367807.7
10998,1000,Company_1000,Consumer Goods,Latin America,2024,3183.2,15.7,6960.1,-6.8,52.0,72.4,40.5,42.9,441310.9,551638.6,2206554.3
10999,1000,Company_1000,Consumer Goods,Latin America,2025,3173.5,17.5,5666.1,-0.3,51.8,74.5,39.7,41.4,430741.6,538427.0,2153707.9


In [7]:
df_raw.info()

<class 'pandas.DataFrame'>
RangeIndex: 11000 entries, 0 to 10999
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CompanyID          11000 non-null  int64  
 1   CompanyName        11000 non-null  str    
 2   Industry           11000 non-null  str    
 3   Region             11000 non-null  str    
 4   Year               11000 non-null  int64  
 5   Revenue            11000 non-null  float64
 6   ProfitMargin       11000 non-null  float64
 7   MarketCap          11000 non-null  float64
 8   GrowthRate         10000 non-null  float64
 9   ESG_Overall        11000 non-null  float64
 10  ESG_Environmental  11000 non-null  float64
 11  ESG_Social         11000 non-null  float64
 12  ESG_Governance     11000 non-null  float64
 13  CarbonEmissions    11000 non-null  float64
 14  WaterUsage         11000 non-null  float64
 15  EnergyConsumption  11000 non-null  float64
dtypes: float64(11), int64(2), str(3)


In [8]:
df_raw.isnull().sum()

CompanyID               0
CompanyName             0
Industry                0
Region                  0
Year                    0
Revenue                 0
ProfitMargin            0
MarketCap               0
GrowthRate           1000
ESG_Overall             0
ESG_Environmental       0
ESG_Social              0
ESG_Governance          0
CarbonEmissions         0
WaterUsage              0
EnergyConsumption       0
dtype: int64

In [9]:
df_raw.describe(include="all")

,CompanyID,CompanyName,Industry,Region,Year,Revenue,ProfitMargin,MarketCap,GrowthRate,ESG_Overall,ESG_Environmental,ESG_Social,ESG_Governance,CarbonEmissions,WaterUsage,EnergyConsumption
count,11000.000000,11000,11000,11000,11000.000000,11000.000000,11000.000000,11000.000000,10000.000000,11000.000000,11000.000000,11000.000000,11000.000000,1.100000e+04,1.100000e+04,1.100000e+04
unique,NaN,1000,9,7,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
top,NaN,Company_1,Healthcare,Asia,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
freq,NaN,11,1331,1672,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
mean,500.500000,NaN,NaN,NaN,2020.000000,4670.850591,10.900455,13380.622236,4.830370,54.615273,56.416991,55.660582,51.767655,1.271462e+06,5.600442e+05,1.165839e+07
std,288.688113,NaN,NaN,NaN,3.162421,9969.954369,8.758711,39922.870373,9.424787,15.893937,26.767233,23.356152,25.323370,5.067760e+06,1.565686e+06,5.095836e+07
min,1.000000,NaN,NaN,NaN,2015.000000,35.900000,-20.000000,1.800000,-36.000000,6.300000,0.000000,0.000000,0.000000,2.042200e+03,1.021100e+03,5.105500e+03
25%,250.750000,NaN,NaN,NaN,2017.000000,938.775000,5.300000,1098.525000,-1.325000,44.100000,34.700000,37.600000,30.775000,1.228530e+05,6.488467e+04,3.069161e+05
50%,500.500000,NaN,NaN,NaN,2020.000000,1902.300000,10.500000,3096.450000,4.900000,54.600000,55.600000,55.150000,52.100000,2.920734e+05,2.038805e+05,1.221745e+06
75%,750.250000,NaN,NaN,NaN,2023.000000,4342.625000,16.300000,9995.500000,11.000000,65.600000,79.000000,73.800000,73.000000,7.407311e+05,5.251880e+05,5.616437e+06


In [10]:
for col in ["Region", "Industry", "Year"]:
    print(f"{col} : {df_raw[col].unique()} \n")

Region : <StringArray>
['Latin America',          'Asia',        'Africa',   'Middle East',
       'Oceania',        'Europe', 'North America']
Length: 7, dtype: str 

Industry : <StringArray>
[        'Retail', 'Transportation',     'Technology',        'Finance',
     'Healthcare',         'Energy', 'Consumer Goods',      'Utilities',
  'Manufacturing']
Length: 9, dtype: str 

Year : [2015 2016 2017 2018 2019 2020 2021 2022 2023 2024 2025] 



In [11]:
for col in ["Region", "Industry"]:
    print(df_raw[col].value_counts())

Region
Asia             1672
Oceania          1661
Middle East      1617
Europe           1540
North America    1540
Latin America    1507
Africa           1463
Name: count, dtype: int64
Industry
Healthcare        1331
Transportation    1287
Manufacturing     1287
Consumer Goods    1276
Finance           1243
Energy            1188
Utilities         1177
Retail            1166
Technology        1045
Name: count, dtype: int64


In [12]:
df_raw["CompanyID"].nunique()

1000

In [13]:
df_raw.duplicated().sum()

np.int64(0)

In [14]:
for col in ["CarbonEmissions","Revenue","WaterUsage"]:
    print(f"{col} < 0 : {(df_raw[col] < 0).sum()}")

print(f"ESG_Overall > 100: {(df_raw['ESG_Overall'] > 100).sum()}")
print(f"ESG_Overall < 0: {(df_raw['ESG_Overall'] < 0).sum()}")

CarbonEmissions < 0 : 0
Revenue < 0 : 0
WaterUsage < 0 : 0
ESG_Overall > 100: 0
ESG_Overall < 0: 0


In [15]:
df_raw[df_raw["ESG_Environmental"] == 0]

,CompanyID,CompanyName,Industry,Region,Year,Revenue,ProfitMargin,MarketCap,GrowthRate,ESG_Overall,ESG_Environmental,ESG_Social,ESG_Governance,CarbonEmissions,WaterUsage,EnergyConsumption
2025,185,Company_185,Transportation,Latin America,2016,340.4,-1.6,1281.9,-19.6,13.2,0.0,21.1,18.4,245118.0,40853.0,817060.1
2026,185,Company_185,Transportation,Latin America,2017,348.6,2.2,147.6,2.4,14.0,0.0,19.6,22.5,250994.2,41832.4,836647.5
2039,186,Company_186,Manufacturing,Middle East,2019,3384.6,13.7,4245.0,-3.8,23.6,0.0,64.1,6.6,1218442.8,1624590.4,20307380.2
3411,311,Company_311,Consumer Goods,Middle East,2016,5933.8,16.9,19235.8,5.1,6.3,0.0,14.5,4.4,1424114.4,1780143.0,7120571.9
5028,458,Company_458,Transportation,Asia,2016,2165.0,6.0,2094.4,-3.8,23.7,0.0,44.5,26.6,1558814.8,259802.5,5196049.2


In [16]:
print((df_raw["ESG_Environmental"] == 0).sum())
print(df_raw[df_raw["ESG_Environmental"] == 0]["CompanyID"].unique())

5
[185 186 311 458]


In [17]:
df_raw["ESG_check"] = (
    df_raw["ESG_Environmental"] + df_raw["ESG_Social"] + df_raw["ESG_Governance"]
) / 3
df_raw["ESG_check"]

0        57.000000
1        56.733333
2        56.466667
3        58.000000
4        56.600000
           ...    
10995    52.200000
10996    51.900000
10997    52.900000
10998    51.933333
10999    51.866667
Name: ESG_check, Length: 11000, dtype: float64

In [18]:
diff_check = (df_raw["ESG_Overall"] - df_raw["ESG_check"]).abs()
print(f"Max Abweichung : {diff_check.max():.3f}")
print(f"Mean Abweichung : {diff_check.mean():.3f}")

Max Abweichung : 0.067
Mean Abweichung : 0.026


### EDA Overview

11.000 rows x 16 cols, covering 1000 unique companies across 9 industries and 7 regions over the timespan of 2015 to 2025

Data Quality:
- No duplicate rows
- GrowthRate shows NaNs for all 1000 companies in 2015 -> First Year, therefore no calculation possible
- No negative Values in CarbonEmissions, Revenue and WaterUsage
- ESG_Overall stays in a valid range across all rows
- 5 rows show ESG Environmental scores of 0.0 across 4 companies, which will be treated as valid ectreme values in the context of a synthetic dataset

Distribution:
- ESG Overall: mean = 54.6 ; range 6.3 - 98.8

## Data Cleaning

In [19]:
df = df_raw.copy()

In [20]:
df.drop(columns=["ESG_check"], inplace=True)

for col in ["Industry","Region","CompanyName"]:
    df[col] = df[col].astype("category")

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 11000 entries, 0 to 10999
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   CompanyID          11000 non-null  int64   
 1   CompanyName        11000 non-null  category
 2   Industry           11000 non-null  category
 3   Region             11000 non-null  category
 4   Year               11000 non-null  int64   
 5   Revenue            11000 non-null  float64 
 6   ProfitMargin       11000 non-null  float64 
 7   MarketCap          11000 non-null  float64 
 8   GrowthRate         10000 non-null  float64 
 9   ESG_Overall        11000 non-null  float64 
 10  ESG_Environmental  11000 non-null  float64 
 11  ESG_Social         11000 non-null  float64 
 12  ESG_Governance     11000 non-null  float64 
 13  CarbonEmissions    11000 non-null  float64 
 14  WaterUsage         11000 non-null  float64 
 15  EnergyConsumption  11000 non-null  float64 
dtypes: category(3),

## Company Selection

In [21]:
eu_com = df[
    (df["Region"] == "Europe") & 
    (df["Industry"] == "Manufacturing") & 
    (df["Year"] == 2025)
].copy()

eu_com["ESG_rank"] = eu_com["ESG_Overall"].rank(ascending=False)
eu_com["CO2_rank"] = eu_com["CarbonEmissions"].rank(ascending=True)

eu_com[["CompanyID", "ESG_Overall", "CarbonEmissions", "Revenue", "ESG_rank", "CO2_rank"]].sort_values("ESG_rank")

,CompanyID,ESG_Overall,CarbonEmissions,Revenue,ESG_rank,CO2_rank
10020,911,86.2,65701.7,438.0,1.0,1.0
8359,760,80.3,509854.7,2198.5,2.0,10.0
5576,507,75.6,153408.4,1012.3,3.0,5.0
7226,657,72.7,127784.9,492.6,4.0,3.0
1165,106,69.4,1091356.5,3263.0,5.0,15.0
9954,905,68.3,126172.8,558.7,6.0,2.0
3299,300,65.7,296714.7,1179.3,7.0,7.0
1000,91,65.4,702579.0,2792.0,8.0,12.0
8524,775,61.7,139122.5,514.8,9.0,4.0
10933,994,59.4,740965.2,2372.4,10.0,13.0


In [22]:
eu_trend = df[
    (df["Region"] == "Europe") & 
    (df["Industry"] == "Manufacturing") & 
    (df["Year"].isin([2015, 2025]))
].pivot_table(
    index="CompanyID",
    columns="Year",
    values=["ESG_Overall","CarbonEmissions"]
).round(1)

eu_trend["ESG_delta"] = eu_trend["ESG_Overall"][2025] - eu_trend["ESG_Overall"][2015] #trend absolut
eu_trend["CO2_delta"] = (
    (eu_trend["CarbonEmissions"][2025] - eu_trend["CarbonEmissions"][2015]) / eu_trend["CarbonEmissions"][2015] * 100 #prozentual
).round(1)

eu_trend.sort_values("ESG_delta", ascending=False)

CarbonEmissions            ESG_Overall       ESG_delta CO2_delta
Year                 2015       2025        2015  2025                    
CompanyID                                                                 
300              569254.8   296714.7        51.8  65.7      13.9     -47.9
224              474632.1   821696.0        39.8  50.2      10.4      73.1
775               55577.1   139122.5        52.4  61.7       9.3     150.3
874              314256.3   432359.2        46.7  55.8       9.1      37.6
657               96600.2   127784.9        63.9  72.7       8.8      32.3
994              355070.0   740965.2        52.8  59.4       6.6     108.7
507              101754.6   153408.4        69.1  75.6       6.5      50.8
911               66563.2    65701.7        79.8  86.2       6.4      -1.3
760              602280.6   509854.7        74.1  80.3       6.2     -15.3
140              247267.5   320881.1        34.0  39.9       5.9      29.8
554              266378.9   566442.3        46.5  51.4       4.9     112.6
106              570002.8  1091356.5        66.1  69.4       3.3      91.5
91              1072422.2   702579.0        62.7  65.4       2.7     -34.5
905              113188.7   126172.8        68.0  68.3       0.3      11.5
773              160715.9   157317.2        58.9  55.3      -3.6      -2.1

### Company selection decision:

Company_300 is selected based on a two-dimensional ranking across all 15 European manufacturers. Company_300 shows both a strong reduction in CarbonEmissions as well as a significant increase in ESG Overall scoring.

Among the European manufacturers Company_300 is the only entity showing both an increase in ESG scoring and reduction in CarbonEmissions, which is in line with the CSRDs double materiality principle.
Other top ESG improvers for example Company_224 show rising Emissions which would raise greenwashing concerns under ESRS E1 disclosure requirements.

Company_300 will serve as Lauschke GmbH for this ESG performance analysis.

## Lauschke GmbH dataset

In [23]:
df_l = df[df["CompanyID"] == 300].copy().reset_index(drop=True)

df_eu = df[
    (df["Region"] == "Europe") & 
    (df["Industry"] == "Manufacturing")
].copy()

df_l

,CompanyID,CompanyName,Industry,Region,Year,Revenue,ProfitMargin,MarketCap,GrowthRate,ESG_Overall,ESG_Environmental,ESG_Social,ESG_Governance,CarbonEmissions,WaterUsage,EnergyConsumption
0,300,Company_300,Manufacturing,Europe,2015,2030.0,2.5,1054.2,NaN,51.8,37.9,53.6,64.1,569254.8,759006.3,9487579.2
1,300,Company_300,Manufacturing,Europe,2016,2008.7,2.2,857.7,-1.1,53.0,40.7,53.6,64.6,551293.1,735057.5,9188218.8
2,300,Company_300,Manufacturing,Europe,2017,1852.9,0.7,248.3,-7.8,52.8,39.3,54.0,65.1,514255.6,685674.2,8570927.0
3,300,Company_300,Manufacturing,Europe,2018,1567.0,0.6,162.6,-15.4,55.2,39.6,59.5,66.5,433770.3,578360.4,7229504.9
4,300,Company_300,Manufacturing,Europe,2019,1406.1,-0.8,5826.7,-10.3,56.6,40.7,61.8,67.3,386137.3,514849.7,6435621.6
5,300,Company_300,Manufacturing,Europe,2020,1164.3,0.1,31.8,-17.2,57.9,43.1,64.7,65.8,313750.3,418333.7,5229171.8
6,300,Company_300,Manufacturing,Europe,2021,1207.1,4.9,902.0,3.7,60.0,46.4,65.6,67.9,316858.4,422477.9,5280973.1
7,300,Company_300,Manufacturing,Europe,2022,1184.4,4.2,463.6,-1.9,60.8,45.4,69.8,67.3,313586.7,418115.6,5226444.7
8,300,Company_300,Manufacturing,Europe,2023,1166.3,6.6,635.9,-1.5,63.2,46.3,70.8,72.5,306368.2,408490.9,5106136.5
9,300,Company_300,Manufacturing,Europe,2024,1093.0,6.6,1220.1,-6.3,63.8,47.0,73.4,71.1,285524.7,380699.6,4758745.1


In [25]:
df_l.to_csv(proc_data_ordner / "lauschke_gmbh.csv", index=False)
df_eu.to_csv(proc_data_ordner / "benchmark.csv", index=False)

print("Done")

Done


Two Datasets saved:
lauschke_gmbh.csv -> Data of Company_300
benchmark.csv -> All European manufacturers for benchmarking